src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala

```diff
@@ -705,12 +705,38 @@ private[snowflake] class SnowflakeSQLStatement(
   def isEmpty: Boolean = list.isEmpty
 
   def execute(
-    bindVariableEnabled: Boolean
-  )(implicit conn: Connection): ResultSet =
-    if (bindVariableEnabled) executeWithBindVariable(conn)
-    else executeWithoutBindVariable(conn)
+               bindVariableEnabled: Boolean
+             )(implicit conn: Connection): ResultSet = {
+    val (rs, queryID, errorThrow) =
+      if (bindVariableEnabled) {
+        executeWithBindVariable(conn)
+      } else {
+        executeWithoutBindVariable(conn)
+      }
+
+    if (errorThrow.isDefined) {
+      log.warn(
+        s"""${SnowflakeResultSetRDD.MASTER_LOG_PREFIX}: Hit error
+           | for query ID: $queryID error message is: ${errorThrow.get.getMessage}
+           |""".stripMargin.filter(_ >= ' '))
+      // Rethrow the exception
+      throw errorThrow.get
+    }
+    rs
+  }
+
+  private[snowflake] def executeWithoutThrow(
+                                              bindVariableEnabled: Boolean
+                                            )(implicit conn: Connection)
+  : (ResultSet, String, Option[Throwable]) =
+    if (bindVariableEnabled) {
+      executeWithBindVariable(conn)
+    } else {
+      executeWithoutBindVariable(conn)
+    }
 
-  private def executeWithoutBindVariable(conn: Connection): ResultSet = {
+  private def executeWithoutBindVariable(conn: Connection)
+  : (ResultSet, String, Option[Throwable]) = {
```

Comment:
Create a case class for this return type.

Reply:
Also, this is not a good way to write this.
```
case class ResultInfo(queryID: String, result: Option[ResultSet])
private def executeWithoutBindVariable(conn): ResultInfo
```
And use Try/Success/Failure to deal with exceptions.
```
def execute() = {
  val executeResult = Try(...)
}
```

Reply:
Or better yet, just use try/catch.

Reply:
The reason to work in this way to retrieve the queryID for query fail.
For successful query, we can get the queryID from resultSet. For failed query, we have to get the queryID from the Statement. But the statement is created in side of executeWithoutBindVariable()/executeWithBindVariable().
To get the queryID and not to introduce duplicate code, I change the internal function to return the Exception and the caller processes the exception.

But I agree it is good idea to use CASE CLASS for the return type. I plan to use ExecuteResult.